pkg/operator/ceph/csi/spec.go

```diff
@@ -683,3 +736,115 @@ func (r *ReconcileCSI) validateCSIVersion(ownerInfo *k8sutil.OwnerInfo) (*CephCS
+	for _, multusCluster := range r.multusClusters {
+		for _, driver := range enabledDrivers {
+			cephPluginHolder, err := templateToDaemonSet("cephpluginholder", driver.holderTemplate, tp)
+			if err != nil {
+				return errors.Wrapf(err, "failed to load ceph %q plugin holder template", driver.fullName)
+			}
+
+			// DO NOT set owner reference on ceph plugin holder daemonset, this DS must never restart unless
+			// the entire node is rebooted
+			holderPluginTolerations := getToleration(r.opConfig.Parameters, driver.toleration, pluginTolerations)
+			holderPluginNodeAffinity := getNodeAffinity(r.opConfig.Parameters, driver.nodeAffinity, pluginNodeAffinity)
+
+			// apply driver's plugin tolerations and node affinity
+			applyToPodSpec(&cephPluginHolder.Spec.Template.Spec, holderPluginNodeAffinity, holderPluginTolerations)
+
+			// apply resource request and limit to *** plugin containers
+			// TODO: expose resources or inherit? the holder needs lesser resources...
+			applyResourcesToContainers(r.opConfig.Parameters, cephFSPluginResource, &cephPluginHolder.Spec.Template.Spec)
+
+			// Append the CEPH_CLUSTER_NAMESPACE env var so that the main container can use it to create the network
+			// namespace symlink to the Kubelet plugin directory
+			cephPluginHolder.Spec.Template.Spec.Containers[0].Env = append(
+				cephPluginHolder.Spec.Template.Spec.Containers[0].Env,
+				corev1.EnvVar{
+					Name:  "CEPH_CLUSTER_NAMESPACE",
+					Value: multusCluster.cluster.Namespace,
+				},
+			)
+
+			// Append the driver name so that the symlink file goes into the right location on the
+			// kubelet plugin directory
+			cephPluginHolder.Spec.Template.Spec.Containers[0].Env = append(
+				cephPluginHolder.Spec.Template.Spec.Containers[0].Env,
+				corev1.EnvVar{
+					Name:  "ROOK_CEPH_CSI_DRIVER_NAME",
+					Value: driver.fullName,
+				},
+			)
+
+			// Make the DS name unique per Ceph cluster
+			cephPluginHolder.Name = fmt.Sprintf("%s-%s", cephPluginHolder.Name, multusCluster.cluster.Name)
+			cephPluginHolder.Spec.Template.Name = fmt.Sprintf("%s-%s", cephPluginHolder.Spec.Template.Name, multusCluster.cluster.Name)
+			cephPluginHolder.Spec.Template.Spec.Containers[0].Name = fmt.Sprintf("%s-%s", cephPluginHolder.Spec.Template.Spec.Containers[0].Name, multusCluster.cluster.Name)
+
+			// Add default labels
+			k8sutil.AddRookVersionLabelToDaemonSet(cephPluginHolder)
+
+			// Apply Multus annotations to daemonset spec
+			err = k8sutil.ApplyMultus(multusCluster.cluster.Spec.Network, &cephPluginHolder.Spec.Template.ObjectMeta)
+			if err != nil {
+				return errors.Wrapf(err, "failed to apply multus configuration for holder %q in cluster %q", cephPluginHolder.Name, multusCluster.cluster.Namespace)
+			}
+
+			// Finally create the DaemonSet
+			err = k8sutil.CreateDaemonSet(r.opManagerContext, r.opConfig.OperatorNamespace, r.context.Clientset, cephPluginHolder)
```

Comment:
internally `CreateDaemonSet` calls update function also, will it be a problematic?

Reply:
I'm not sure I understand the concern, can you clarify?

Reply:
https://github.com/rook/rook/blob/d5643b439d7def76b0eb2c6052ebbba84244a035/pkg/operator/k8sutil/daemonset.go#L34 am thinking can be cause any problem as it also update the ds holder template?

Reply:
I think I see what you mean, this holder should never be updated, not even by our controller unless there is a security issue. Calling update is probably not a good idea even though most of the time it should be idempotent. I'll change that.